buzz_el/entity_matcher/entity_matcher.py

```diff
@@ -0,0 +1,158 @@
+from typing import Dict, Iterable, List, Optional
+
+from spacy.language import Language
+from spacy.pipeline import SpanRuler
+from spacy.tokens import Doc
+
+from ..graph_loader import GraphLoader
+
+
+class EntityMatcher:
+    """
+    A class to construct a spaCy span ruler from a RDF graph.
+
+    Parameters
+    ----------
+    graph_loader : GraphLoader
+        An instance of the GraphLoader class for loading knowledge graph data.
+    spacy_model : Language
+        The spaCy language to use when constructing the spaCy span ruler.
+    span_ruler_config : Optional[Dict], optional
+        Configuration for the spaCy span ruler, by default {"phrase_matcher_attr": "LOWER"}.
+        See <https://spacy.io/api/spanruler#config>
+
+    Attributes
+    ----------
+    spacy_model : Language
+        The spaCy language to use when constructing the spaCy span ruler.
+    graph_loader : GraphLoader
+        An instance of the GraphLoader class for loading knowledge graph data.
+    _span_ruler_config : Dict
+        Configuration for the spaCy span ruler, by default {"phrase_matcher_attr": "LOWER"}.
+        See: <https://spacy.io/api/spanruler#config>
+    spacy_component : SpanRuler
+        The spaCy span ruler component for the knowledge graph entity matching.
+    """
+
+    def __init__(
+        self,
+        graph_loader: GraphLoader,
+        spacy_model: Language,
+        span_ruler_config: Optional[Dict] = None,
+    ) -> None:
+        """
+        Initialiser for the entity matcher.
+
+        Parameters
+        ----------
+        graph_loader : GraphLoader
+            An instance of the GraphLoader class for loading knowledge graph data.
+        spacy_model : Language
+            The spaCy language to use when constructing the spaCy span ruler.
+        span_ruler_config : Optional[Dict], optional
+            Configuration for the spaCy span ruler, by default {"phrase_matcher_attr": "LOWER"}.
+            See: <https://spacy.io/api/spanruler#config>
+        """
+        self.spacy_model = spacy_model
+        self.graph_loader = graph_loader
+
+        if span_ruler_config is None:
+            self._span_ruler_config = {"phrase_matcher_attr": "LOWER"}
+        else:
+            self._span_ruler_config = span_ruler_config
+
+        self.build_entity_ruler()  # set self.spacy_component
+
+    def __call__(self, doc: Doc) -> None:
```

Comment:
If I understand correctly, by default you can use the call on a single doc and if you want to apply it to the whole corpus it's the pipe function?

Reply:
Yes, I mapped how spaCy is working. OR at least I tried ;)